worlds/saving_princess/Rules.py

```diff
@@ -0,0 +1,110 @@
+from BaseClasses import CollectionState, Location
+from worlds.generic.Rules import set_rule
+
+from .Constants import *
+
+
+def set_rules(world: "SavingPrincessWorld"):
+    def get_location(name: str) -> Location:
+        return world.multiworld.get_location(name, world.player)
+
+    def get_region_entrance(name: str) -> Location:
+        return world.multiworld.get_entrance(f"{name} entrance", world.player)
+
+    def can_hover(state: CollectionState) -> bool:
+        # portia can hover if she has a weapon other than the powered blaster and 4 reload speed upgrades
+        return (
+                state.has(ITEM_RELOAD_SPEED, world.player, 4)
+                and (
+                        state.has(ITEM_WEAPON_FIRE, world.player)
+                        or state.has(ITEM_WEAPON_ICE, world.player)
+                        or state.has(ITEM_WEAPON_VOLT, world.player)
+                )
+        )
+
+    # guarantees that the player will have some upgrades before having to face the area bosses, except for cave
+    def nice_check(state: CollectionState) -> bool:
+        return (
+                state.has(ITEM_MAX_HEALTH, world.player, 1)
+                and state.has(ITEM_MAX_AMMO, world.player, 1)
+                and state.has(ITEM_RELOAD_SPEED, world.player, 2)
+        )
+
+    # same as above, but for the final area
+    def super_nice_check(state: CollectionState) -> bool:
+        return (
+                state.has(ITEM_MAX_HEALTH, world.player, 2)
+                and state.has(ITEM_MAX_AMMO, world.player, 2)
+                and state.has(ITEM_RELOAD_SPEED, world.player, 4)
+                and state.has(ITEM_WEAPON_CHARGE, world.player)
+                # at least one special weapon, other than powered blaster
+                and (
+                        state.has(ITEM_WEAPON_FIRE, world.player)
+                        or state.has(ITEM_WEAPON_ICE, world.player)
+                        or state.has(ITEM_WEAPON_VOLT, world.player)
+                )
+        )
+
+    def is_gate_unlocked(state: CollectionState) -> bool:
+        # the gate unlocks with all 4 boss keys, although this only applies to extended pool
+        if world.is_pool_expanded:
+            # in expanded, the final area requires all the boss keys
+            return (
+                    state.has(EP_ITEM_GUARD_GONE, world.player)
+                    and state.has(EP_ITEM_CLIFF_GONE, world.player)
+                    and state.has(EP_ITEM_ACE_GONE, world.player)
+                    and state.has(EP_ITEM_SNAKE_GONE, world.player)
```

Comment:
```suggestion
                    state.has_all({EP_ITEM_GUARD_GONE, EP_ITEM_CLIFF_GONE, EP_ITEM_ACE_GONE, EP_ITEM_SNAKE_GONE}, world.player)
```

You can use state.has_all to do the same thing you had here